arangod/Cache/PlainBucket.cpp

```diff
@@ -95,57 +86,96 @@ CachedValue* PlainBucket::find(std::uint32_t hash, void const* key,
 // requires there to be an open slot, otherwise will not be inserted
 void PlainBucket::insert(std::uint32_t hash, CachedValue* value) noexcept {
   TRI_ASSERT(isLocked());
-  for (std::size_t i = 0; i < slotsData; i++) {
-    if (_cachedHashes[i] == 0) {
-      // found an empty slot
-      _cachedHashes[i] = hash;
-      _cachedData[i] = value;
-      if (i != 0) {
-        moveSlot(i, true);
-      }
-      return;
+  if (_slotsUsed < kSlotsData) {
+    // found an empty slot
+    TRI_ASSERT(_cachedData[_slotsUsed] == nullptr);
+    _cachedHashes[_slotsUsed] = hash;
+    _cachedData[_slotsUsed] = value;
+    if (_slotsUsed != 0) {
+      moveSlot(_slotsUsed);
     }
+    ++_slotsUsed;
+    TRI_ASSERT(_slotsUsed <= kSlotsData);
+    checkInvariants();
   }
 }
 
 template<typename Hasher>
 CachedValue* PlainBucket::remove(std::uint32_t hash, void const* key,
                                  std::size_t keySize) noexcept {
   TRI_ASSERT(isLocked());
-  CachedValue* value = find<Hasher>(hash, key, keySize, false);
-  if (value != nullptr) {
-    evict(value, false);
+  CachedValue* result = nullptr;
+
+  for (std::size_t i = 0; i < _slotsUsed; i++) {
+    if (_cachedHashes[i] == hash &&
+        Hasher::sameKey(_cachedData[i]->key(), _cachedData[i]->keySize(), key,
+                        keySize)) {
+      result = _cachedData[i];
+      _cachedHashes[i] = _cachedHashes[_slotsUsed - 1];
+      _cachedData[i] = _cachedData[_slotsUsed - 1];
+      _cachedHashes[_slotsUsed - 1] = 0;
+      _cachedData[_slotsUsed - 1] = nullptr;
+      TRI_ASSERT(_slotsUsed > 0);
+      --_slotsUsed;
+      checkInvariants();
+      break;
+    }
   }
 
-  return value;
+  return result;
 }
 
-CachedValue* PlainBucket::evictionCandidate(
-    bool ignoreRefCount) const noexcept {
+std::uint64_t PlainBucket::evictCandidate() noexcept {
   TRI_ASSERT(isLocked());
-  for (std::size_t i = 0; i < slotsData; i++) {
-    std::size_t slot = slotsData - (i + 1);
-    if (_cachedData[slot] == nullptr) {
+  std::size_t slot = _slotsUsed;
+  while (slot-- > 0) {
+    TRI_ASSERT(_cachedData[slot] != nullptr);
+    if (!_cachedData[slot]->isFreeable()) {
       continue;
     }
-    if (ignoreRefCount || _cachedData[slot]->isFreeable()) {
+
+    std::uint64_t size = _cachedData[slot]->size();
+    // evict value. we checked that it is freeable
+    delete _cachedData[slot];
+    _cachedHashes[slot] = _cachedHashes[_slotsUsed - 1];
+    _cachedData[slot] = _cachedData[_slotsUsed - 1];
+    _cachedHashes[_slotsUsed - 1] = 0;
+    _cachedData[_slotsUsed - 1] = nullptr;
+    TRI_ASSERT(_slotsUsed > 0);
+    --_slotsUsed;
+    checkInvariants();
```

Comment:
This is the same code as in remove. Can we split this into a separate helper function?